chore(curriculum): add football team cards lab to frontend cert (#56129)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-football-team-cards/66e7ee20b79186306fc12da5.md

```

The `headCoach` property should be a string.

```js
assert.isString(footballTeam.headCoach);
```

The `players` property should be an array of at least four objects, each object should have the keys `name`, `position`, `isCaptain`.

```js
assert.isArray(footballTeam.players);
assert.isAtLeast(footballTeam.players.length, 4);
footballTeam.players.forEach(player => {
  assert.isObject(player);
  assert.containsAllKeys(player, ['name', 'position', 'isCaptain']);
})
```

The `name` property should have value of a string.

```js
footballTeam.players.forEach(({name}) => assert.isString(name));
```

The `position` property should have one of values `"forward"`, `"midfielder"`, `"defender"`, or `"goalkeeper"`.

```js
footballTeam.players.forEach(({position}) => {
  assert.isString(position);
  assert.oneOf(position, ["forward", "midfielder", "defender", "goalkeeper"]);
});
```

The `isCaptain` property should have value of a boolean, and there should be only one captain.

```js
footballTeam.players.forEach(({isCaptain}) => assert.isBoolean(isCaptain));
const listOfCaptains = footballTeam.players.filter(({isCaptain}) => isCaptain);
assert.lengthOf(listOfCaptains, 1);
```

> You should display the values from the `footballTeam` object on the page.

Reviewer: ```suggestion
You should display the `coach`, `team` and `year` values from the `footballTeam` object in the HTML elements with the `id` values of `head-coach`, `team` and `year`. 
```

---
id: 66e7ee20b79186306fc12da5
title: Build Football Team Cards
challengeType: 14 
dashedName: lab-football-team-cards

> # demoType can either be 'onClick' or 'onLoad'. If the project or lab doesn't have a preview, delete the property

Reviewer 2: this comment can be removed

demoType: onClick
---

# --description--

> In this lab, you will write the functionality of adding data to the page from a JavaScript object, and of filtering dinamically these datas.

Reviewer 2: ```suggestion
In this lab, you will build a set of football team cards. The user should be able to use the dropdown menu and filter between the different players based on their positions.
```

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

**User Stories:**

> 1. You should create a `footballTeam` object, this object should have properties `team`, `year`, `headCoach`, `players`.

Reviewer 2: ```suggestion
1. You should create a `footballTeam` object with the following properties: `team`, `year`, `headCoach`, `players`.
```

1. The `team` property should contain the name of the team as a string.
1. The `year` property should contain the year as a number.
1. The `headCoach` property should contain the name of the head coach as a string.
1. The `players` property should be an array containing at least four elements.
1. Each element in the `players` array should be an object with properties `name`, `position`, `isCaptain`.
1. The `name` property should contain the name of the player as a string.
1. The `position` property should have one of the following values: `"forward"`, `"midfielder"`, `"defender"`, or `"goalkeeper"`.

> 1. The `isCaptain` property should have value of a boolean, and it should be `true` only once between all the players.

Reviewer 2: ```suggestion
1. The `isCaptain` property should have value of a boolean. One of the players should have their `isCaptain` property set to `true`.
```

> 1. You should display the values from the object on the page.

Reviewer 2: I would be more specific here

```suggestion
1. You should display the `coach`, `team` and `year` values on the page. These values should be displayed in the HTML elements with the `id` values of `head-coach`, `team` and `year`. 
```

> 1. You should implent a filtering function: when the dropdown menu is used to select one of the positions, only players of that position should be shown.

Reviewer 2: ```suggestion
1. When the dropdown menu is used to select one of the positions, only players of that position should be shown. If the `"All Players"` option is selected, then all of the players should display on the page.
```

# --hints--

You should have a `footballTeam` variable.

```js
assert.exists(footballTeam);
```

The `footballTeam` variable should be an object with four properties: `team`, `year`, `headCoach` and `players`.

```js
assert.isObject(footballTeam);
assert.hasAllKeys(footballTeam, ['team', 'year', 'headCoach', 'players']);

> // it could be containsAllKeys if we want to give freedom to have more than these keys

Reviewer 2: yeah, I think that should be fine to give them the freedom to have additional keys if they want. 
so, as long as they have the 4 keys we are looking for them it is fine if they want to add more
